### PR TITLE
Adds the cancellable ban events and implements them in the server, resolves #678

### LINKED ISSUE #678: please, Add cancellable (un)ban(ip) Events
Hello. I think everything is clear from the title.
I don't understand why this is not done in a Spigot and I will be glad if at least you make it.

@Zbob750, so what do you think about my proposal to add beautiful explosions?

### PR DESCRIPTION
**Pull request by dignissimus**

### CODE REVIEW

File: Spigot-API-Patches/0067-Added-ban-events.patch [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
++        return cancel;
++    }
++
++    @Override
++    public void setCancelled(boolean cancel) {
++        this.cancel = cancel;
++    }
++
++    @Override
++    public HandlerList getHandlers() {
```
Comment (kashike) [inline Spigot-API-Patches/0067-Added-ban-events.patch]: Missing `static` version of this method.

File: Spigot-API-Patches/0067-Added-ban-events.patch [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
++ * Called when a player is banned
++ */
++public abstract class BanEvent extends ServerEvent implements Cancellable {
++    private final Date created;
++    private Date expiration;
++    private String source;
++    private String reason;
++    private String target;
++    private boolean cancel = false;
++    private final HandlerList handlers = new HandlerList();
```
Comment (kashike) [inline Spigot-API-Patches/0067-Added-ban-events.patch]: Should be `static`.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0067-Added-ban-events.patch [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
++
++    @Override
++    public void setCancelled(boolean cancel) {
++        this.cancel = cancel;
++    }
++    @Override
++    public HandlerList getHandlers(){
++    	return handlers;
++    }
++    @Override
```
Comment (kashike) [inline Spigot-API-Patches/0067-Added-ban-events.patch]: ...

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0067-Added-ban-events.patch [hunk 4]
```diff
@@ -56,7 +56,7 @@ index 00000000..46804b2f
 +    public HandlerList getHandlers(){
```
Comment (mikroskeem) [inline Spigot-API-Patches/0067-Added-ban-events.patch]: whitespace? `() {`

**Review by mikroskeem — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
++        // Paper end
+         this.h();
+         return (V) this.d.get(this.a(k0)); // CraftBukkit - fix decompile error
+     }
+@@ -112,7 +114,8 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+         return this.d.size() < 1;
+     }
+ 
+-    protected String a(K k0) {
++    //Paper Start - fix decompile error
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:142]: This doesn't have a Paper end, but neither this nor the get() method even need to use the start/end style. Just do a single line change/comment.

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+     }
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+index 80832f78..05791767 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftIpBanList.java
+@@ -1,19 +1,19 @@
+ package org.bukkit.craftbukkit;
+ 
+-import java.io.IOException;
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:154]: Please clean up this import shuffle. Keep it as it originally was.

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+     @Override
+     public org.bukkit.BanEntry addBan(String target, String reason, Date expires, String source) {
++
+         Validate.notNull(target, "Ban target cannot be null");
+ 
+         IpBanEntry entry = new IpBanEntry(target, new Date(),
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:190]: Missing Paper tags

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch (hunk 6 above)
Comment (aikar) [reply Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:154]: Also should add paper tags

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
++        list.add(newEntry);
+ 
+         try {
+             list.save();
+@@ -50,7 +57,7 @@ public class CraftIpBanList implements org.bukkit.BanList {
+             Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-ips.json, {0}", ex.getMessage());
+         }
+ 
+-        return new CraftIpBanEntry(target, entry, list);
++        return new CraftIpBanEntry(event.getTarget(), newEntry, list);
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:206]: Paper tag

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+     }
+ 
+     @Override
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
+index 84a1f9c8..2ac9db43 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
+@@ -1,21 +1,21 @@
+ package org.bukkit.craftbukkit;
+
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:216]: restore import order to orig, add paper tags

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+ import java.util.logging.Level;
+-import org.bukkit.Bukkit;
+ 
+ public class CraftProfileBanList implements org.bukkit.BanList {
+     private final GameProfileBanList list;
+@@ -54,7 +54,14 @@ public class CraftProfileBanList implements org.bukkit.BanList {
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:248]: Paper tags

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
++        list.add(newEntry);
+ 
+         try {
+             list.save();
+@@ -62,7 +69,7 @@ public class CraftProfileBanList implements org.bukkit.BanList {
+             Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-players.json, {0}", ex.getMessage());
+         }
+ 
+-        return new CraftProfileBanEntry(profile, entry, list);
++        return new CraftProfileBanEntry(newProfile, newEntry, list);
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:265]: Paper tag

File: Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+     }
+diff --git a/src/main/java/net/minecraft/server/IpBanList.java b/src/main/java/net/minecraft/server/IpBanList.java
+index 617d326f..1368c715 100644
+--- a/src/main/java/net/minecraft/server/IpBanList.java
++++ b/src/main/java/net/minecraft/server/IpBanList.java
+@@ -1,6 +1,9 @@
+ package net.minecraft.server;
+ 
+ import com.google.gson.JsonObject;
++import org.bukkit.craftbukkit.CraftIpBanEntry;
```
Comment (aikar) [inline Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch:93]: Paper tags

File: Spigot-API-Patches/0067-Added-ban-events.patch [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.server.ServerEvent;
++
++import java.util.Date;
++
++/**
++ * Called when a player is banned
++ */
++public abstract class BanEvent extends ServerEvent implements Cancellable {
```
Comment (aikar) [inline Spigot-API-Patches/0067-Added-ban-events.patch:25]: I don't see a reason to extend ServerEvent here. Why does that class even exists...

Especially considering cases of this event being triggered may be player oriented, not server.

Which adds next request. Replace getSource and all usages of these strings with the CommandSender instead.  a string is not accurate enough to determine the proper sender (IE: For Command Blocks or Console Sender)

the source string should not be exposed.

I'm also on the fence about target. Id rather Profile oriented bans return an OfflinePlayer, and IP bans give the IP in string or maybe a representative Java Object (InetAddress I think it is?)

**Review by aikar — CHANGES_REQUESTED** [review]

**Comment by electronicboy** [comment]
Closing as OP has shown intents to rework this into another PR.

**Pull request closed without merging**